{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Testing Expasy chat\n",
    "\n",
    "These tests requires to start the expasy chat stack locally with `docker compose -f compose.dev.yml up`\n",
    "\n",
    "Testing our Expasy query helper reusing the example queries that succeeded retrieved from `test_example_queries.ipynb`\n",
    "\n",
    "Potential interesting questions to use for evaluation:\n",
    "- Which are the human genes associated with cancer and their orthologs expressed in the rat brain?\n",
    "- Select the number of proteins for each of the subclasses of ec:1.1.1.- EC (Enzyme Commission) class (from biosoda)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from collections import defaultdict\n",
    "import time\n",
    "from weakref import ref\n",
    "\n",
    "import pandas as pd\n",
    "import requests\n",
    "from dotenv import load_dotenv\n",
    "\n",
    "from sparql_llm.config import get_llm_client, settings\n",
    "from sparql_llm.utils import query_sparql\n",
    "from sparql_llm.validate_sparql import extract_sparql_queries\n",
    "\n",
    "load_dotenv()\n",
    "expasy_api_key = os.getenv(\"EXPASY_API_KEY\")\n",
    "\n",
    "example_queries = [\n",
    "    {\n",
    "        \"question\": \"What is the accession number in uniprot of the human gene LCT? Return only unique protein URIs\",\n",
    "        \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "        \"query\": \"\"\"PREFIX up: <http://purl.uniprot.org/core/>\n",
    "    PREFIX skos: <http://www.w3.org/2004/02/skos/core#>\n",
    "    PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
    "    SELECT DISTINCT ?protein\n",
    "    WHERE{\n",
    "        ?protein a up:Protein .\n",
    "        ?protein up:organism taxon:9606 .\n",
    "        ?protein up:encodedBy ?gene .\n",
    "        ?gene skos:prefLabel \"LCT\" .\n",
    "    }\"\"\",\n",
    "    },\n",
    "    {\n",
    "        # NOTE: The \"mature\" part in the question makes it harder to answer\n",
    "        # \"question\": \"How do I filter for reviewed (mouse) proteins whose mature form carries an N-terminal glycine? Return protein URI and AA sequence\",\n",
    "        \"question\": \"How do I filter for reviewed mouse proteins which carry an N-terminal glycine? Return protein URI and AA sequence\",\n",
    "        \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "        \"query\": \"\"\"PREFIX up: <http://purl.uniprot.org/core/>\n",
    "    PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
    "    PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\n",
    "    SELECT ?protein ?sequence\n",
    "    WHERE\n",
    "    {\n",
    "        ?protein a up:Protein ;\n",
    "            up:organism taxon:10090 ;  # Taxonomy ID for Mus musculus (Mouse)\n",
    "            up:reviewed true ;\n",
    "            up:sequence ?isoform .\n",
    "        ?isoform rdf:value ?sequence .\n",
    "        # Ensure the N-terminal amino acid is Glycine (G)\n",
    "        FILTER (STRSTARTS(?sequence, \"G\"))\n",
    "    }\"\"\",\n",
    "    },\n",
    "    {\n",
    "        \"question\": \"How could I download a table that only includes the Rhea reactions for which there is experimental evidence? Return only the rhea URI\",\n",
    "        \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "        \"query\": \"\"\"PREFIX up: <http://purl.uniprot.org/core/>\n",
    "    PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\n",
    "    SELECT DISTINCT ?rhea\n",
    "    WHERE {\n",
    "    # ECO 269 is experimental evidence\n",
    "    BIND (<http://purl.obolibrary.org/obo/ECO_0000269> as ?evidence)\n",
    "    #GRAPH <http://sparql.uniprot.org/uniprot> {\n",
    "        ?protein up:reviewed true ;\n",
    "        up:annotation ?a ;\n",
    "        up:attribution ?attribution  .\n",
    "\n",
    "        ?a a up:Catalytic_Activity_Annotation ;\n",
    "        up:catalyticActivity ?ca .\n",
    "        ?ca up:catalyzedReaction ?rhea .\n",
    "\n",
    "        [] rdf:subject ?a ;\n",
    "        rdf:predicate up:catalyticActivity ;\n",
    "        rdf:object ?ca ;\n",
    "        up:attribution ?attribution .\n",
    "\n",
    "        ?attribution up:evidence ?evidence .\n",
    "    #}\n",
    "    }\"\"\",\n",
    "    },\n",
    "    {\n",
    "        \"question\": \"Which human proteins are enzymes catalyzing a reaction involving sterols? Return the protein, sterol and reaction URI\",\n",
    "        \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "        \"query\": \"\"\"PREFIX up: <http://purl.uniprot.org/core/>\n",
    "    PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
    "    PREFIX rh: <http://rdf.rhea-db.org/>\n",
    "    PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
    "    PREFIX CHEBI: <http://purl.obolibrary.org/obo/CHEBI_>\n",
    "    SELECT DISTINCT ?protein ?sterol ?reaction\n",
    "    WHERE {\n",
    "    SERVICE <https://sparql.rhea-db.org/sparql> {\n",
    "        ?reaction rdfs:subClassOf rh:Reaction .\n",
    "        ?reaction rh:side/rh:contains/rh:compound ?compound .\n",
    "        ?compound rh:chebi ?sterol .\n",
    "        ?sterol rdfs:subClassOf* CHEBI:15889 .\n",
    "    }\n",
    "    ?protein a up:Protein ;\n",
    "        up:organism taxon:9606 ;\n",
    "        up:annotation/up:catalyticActivity/up:catalyzedReaction ?reaction .\n",
    "    }\"\"\",\n",
    "    },\n",
    "    {\n",
    "        # \"question\": \"Which are the human proteins associated with cancer (which have cancer in their disease label)? Return the unique disease label (?diseaseLabel), and HGNC symbol (?hgncSymbol)\",\n",
    "        \"question\": \"Which are the human proteins associated with cancer? Return distinct ?diseaseLabel and ?hgncSymbol\",\n",
    "        \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "        \"query\": \"\"\"PREFIX skos: <http://www.w3.org/2004/02/skos/core#>\n",
    "    PREFIX up:<http://purl.uniprot.org/core/>\n",
    "    PREFIX taxon:<http://purl.uniprot.org/taxonomy/>\n",
    "    PREFIX rdfs:<http://www.w3.org/2000/01/rdf-schema#>\n",
    "    PREFIX orth:<http://purl.org/net/orth#>\n",
    "    PREFIX dcterms:<http://purl.org/dc/terms/>\n",
    "    PREFIX obo:<http://purl.obolibrary.org/obo/>\n",
    "    PREFIX lscr:<http://purl.org/lscr#>\n",
    "    PREFIX genex:<http://purl.org/genex#>\n",
    "    PREFIX sio: <http://semanticscience.org/resource/>\n",
    "    SELECT DISTINCT ?diseaseLabel ?hgncSymbol\n",
    "    WHERE {\n",
    "        ?humanProtein a up:Protein ;\n",
    "            # up:organism/up:scientificName 'Homo sapiens' ;\n",
    "            up:organism taxon:9606 ;\n",
    "            up:annotation ?annotation ;\n",
    "            rdfs:seeAlso ?hgnc .\n",
    "        ?hgnc up:database <http://purl.uniprot.org/database/HGNC> ;\n",
    "            rdfs:comment ?hgncSymbol .\n",
    "        ?annotation a up:Disease_Annotation ;\n",
    "            up:disease ?disease .\n",
    "        ?disease skos:prefLabel ?diseaseLabel.\n",
    "        FILTER CONTAINS (LCASE(?diseaseLabel), \"cancer\")\n",
    "    }\"\"\",\n",
    "    },\n",
    "    {\n",
    "        \"question\": \"In bgee how can I retrieve the confidence level and false discovery rate of a gene expression? Return distinct ?gene, ?confidence and ?fdr, limit to 10\",\n",
    "        \"endpoint\": \"https://www.bgee.org/sparql/\",\n",
    "        \"query\": \"\"\"PREFIX genex: <http://purl.org/genex#>\n",
    "    PREFIX bgee: <http://bgee.org/#>\n",
    "    PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
    "    SELECT DISTINCT ?gene ?confidence ?fdr\n",
    "    WHERE {\n",
    "        ?expression a genex:Expression ;\n",
    "            genex:hasConfidenceLevel ?confidence ;\n",
    "            genex:hasFDRpvalue ?fdr ;\n",
    "            genex:hasSequenceUnit ?gene .\n",
    "    } LIMIT 10\"\"\",\n",
    "    },\n",
    "    {\n",
    "        # There are no example with xrefEnsembl, so the RAG without validation usually fails\n",
    "        \"question\": \"How can I get the cross-reference to the ensembl protein for the LCT protein in OMA? Return only the distinct ?ensemblURI\",\n",
    "        \"endpoint\": \"https://sparql.omabrowser.org/sparql/\",\n",
    "        \"query\": \"\"\"PREFIX orth: <http://purl.org/net/orth#>\n",
    "    PREFIX lscr: <http://purl.org/lscr#>\n",
    "    PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
    "    SELECT DISTINCT ?ensemblURI\n",
    "    WHERE {\n",
    "        ?protein a orth:Protein ;\n",
    "            rdfs:label 'LCT' ;\n",
    "            lscr:xrefEnsemblProtein ?ensemblURI .\n",
    "    }\"\"\",\n",
    "    },\n",
    "    {\n",
    "        # There are no example with inDataset\n",
    "        \"question\": \"How can I get the URI of a dataset to which an ortholog cluster belongs in OMA? Return orthologCluster, datasetURI and limit to 20\",\n",
    "        \"endpoint\": \"https://sparql.omabrowser.org/sparql/\",\n",
    "        \"query\": \"\"\"PREFIX orth: <http://purl.org/net/orth#>\n",
    "    SELECT DISTINCT ?orthologCluster ?datasetURI\n",
    "    WHERE {\n",
    "        ?orthologCluster a orth:OrthologsCluster ;\n",
    "            orth:inDataset ?datasetURI .\n",
    "    } LIMIT 20\"\"\",\n",
    "    },\n",
    "    {\n",
    "        \"question\": \"Give me the list of strains associated to the Escherichia coli taxon and their name. Return ?taxon, ?strain, ?name, limit to 20\",\n",
    "        \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "        \"query\": \"\"\"PREFIX up: <http://purl.uniprot.org/core/>\n",
    "    PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
    "    PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
    "    SELECT ?taxon ?strain ?name\n",
    "    WHERE {\n",
    "        ?taxon a up:Taxon .\n",
    "        ?taxon rdfs:subClassOf taxon:562 . # Escherichia coli taxon\n",
    "        ?taxon up:strain ?strain .\n",
    "        ?strain up:name ?name .\n",
    "    } LIMIT 20\"\"\",\n",
    "    },\n",
    "    {\n",
    "        \"question\": \"Retrieve all proteins involved in pathways involving glycolysis. Return ?proteinURI, ?proteinLabel, ?pathwayLabel, limit to 20\",\n",
    "        \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "        \"query\": \"\"\"PREFIX up: <http://purl.uniprot.org/core/>\n",
    "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
    "SELECT DISTINCT ?proteinURI ?proteinLabel ?pathwayLabel\n",
    "WHERE {\n",
    "    ?proteinURI a up:Protein ;\n",
    "        up:recommendedName/up:fullName ?proteinLabel ;\n",
    "        up:annotation ?annotation .\n",
    "    ?annotation a up:Pathway_Annotation ;\n",
    "        rdfs:seeAlso ?pathway .\n",
    "    ?pathway rdfs:label ?pathwayLabel .\n",
    "    FILTER(CONTAINS(LCASE(?pathwayLabel), \"glycolysis\"))\n",
    "} LIMIT 20\"\"\",\n",
    "    },\n",
    "    {\n",
    "        \"question\": \"What are the orthologs in rat for protein Q9Y2T1 ? Return ?ratProtein ?ratUniProtXref\",\n",
    "        \"endpoint\": \"https://sparql.omabrowser.org/sparql/\",\n",
    "        \"query\": \"\"\"PREFIX up: <http://purl.uniprot.org/core/>\n",
    "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
    "PREFIX orth: <http://purl.org/net/orth#>\n",
    "PREFIX lscr: <http://purl.org/lscr#>\n",
    "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
    "SELECT DISTINCT ?ratProtein ?ratUniProtXref\n",
    "WHERE {\n",
    "    ?cluster a orth:OrthologsCluster.\n",
    "    ?cluster orth:hasHomologousMember ?node1.\n",
    "    ?cluster orth:hasHomologousMember ?node2.\n",
    "    ?node1 orth:hasHomologousMember* ?humanProtein.\n",
    "    ?node2 orth:hasHomologousMember* ?ratProtein.\n",
    "    ?humanProtein a orth:Protein;\n",
    "        lscr:xrefUniprot <http://purl.uniprot.org/uniprot/Q9Y2T1>.\n",
    "    ?ratProtein a orth:Protein;\n",
    "        orth:organism/obo:RO_0002162/up:scientificName 'Rattus norvegicus';\n",
    "        lscr:xrefUniprot ?ratUniProtXref.\n",
    "    FILTER(?node1 != ?node2)\n",
    "}\"\"\",\n",
    "    },\n",
    "    {\n",
    "        \"question\": \"\"\"Retrieve all proteins that are associated with Alzheimer disease (http://purl.uniprot.org/diseases/3832) and where they are known to be located in the cell. Return ?proteinURI, ?locationInsideCellLabel, ?locationInsideCellUri, limit to 20\"\"\",\n",
    "        \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "        \"query\": \"\"\"PREFIX up: <http://purl.uniprot.org/core/>\n",
    "PREFIX skos: <http://www.w3.org/2004/02/skos/core#>\n",
    "SELECT ?proteinURI ?locationInsideCellLabel ?locationInsideCellUri\n",
    "WHERE {\n",
    "    ?proteinURI a up:Protein ;\n",
    "        up:annotation ?diseaseAnnotation , ?subcellAnnotation .\n",
    "    ?diseaseAnnotation up:disease <http://purl.uniprot.org/diseases/3832> .\n",
    "    ?subcellAnnotation up:locatedIn/up:cellularComponent ?locationInsideCellUri .\n",
    "    ?locationInsideCellUri skos:prefLabel ?locationInsideCellLabel .\n",
    "} LIMIT 20\"\"\",\n",
    "    },\n",
    "    {\n",
    "        \"question\": \"Retrieve all proteins in OMA that are encoded by the TP53 gene and their mnemonics and evidence types from the UniProt database. Return ?proteinOMA ?speciesLabel  ?mnemonic  ?evidenceType ?uniprotURI\",\n",
    "        \"endpoint\": \"https://sparql.omabrowser.org/sparql/\",\n",
    "        \"query\": \"\"\"PREFIX up: <http://purl.uniprot.org/core/>\n",
    "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
    "PREFIX orth: <http://purl.org/net/orth#>\n",
    "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
    "PREFIX lscr: <http://purl.org/lscr#>\n",
    "SELECT DISTINCT ?proteinOMA ?speciesLabel ?mnemonic ?evidenceType ?uniprotURI\n",
    "WHERE {\n",
    "    ?proteinOMA a orth:Protein ;\n",
    "        orth:organism/obo:RO_0002162/up:scientificName ?speciesLabel ;\n",
    "        rdfs:label 'TP53' .\n",
    "    ?proteinOMA lscr:xrefUniprot ?uniprotURI.\n",
    "    SERVICE <http://sparql.uniprot.org/sparql> {\n",
    "        ?uniprotURI up:mnemonic ?mnemonic ;\n",
    "            up:existence/rdfs:label ?evidenceType .\n",
    "    }\n",
    "}\n",
    "\"\"\",\n",
    "    },\n",
    "\n",
    "    # New queries to test:\n",
    "    # Find all proteins linked to arachidonate (CHEBI:32395) and their associated pathways\n",
    "    # List all enzymes that have been experimentally validated and are involved in DNA repair\n",
    "    # Find all proteins that have a mutagenesis annotation affecting their active site\n",
    "\n",
    "\n",
    "    #     {\n",
    "    #         # Way too slow for some reason\n",
    "    #         \"question\": \"Retrieve all proteins that are associated with Alzheimer disease. Return ?proteinURI, ?proteinLabel, ?diseaseURI, ?diseaseLabel, limit to 20\",\n",
    "    #         \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "    #         \"query\": \"\"\"PREFIX up: <http://purl.uniprot.org/core/>\n",
    "    # PREFIX skos: <http://www.w3.org/2004/02/skos/core#>\n",
    "    # PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
    "    # SELECT ?proteinURI ?proteinLabel ?diseaseURI ?diseaseLabel\n",
    "    # WHERE {\n",
    "    #     ?proteinURI a up:Protein ;\n",
    "    #             up:annotation ?diseaseAnnotation ;\n",
    "    #             up:recommendedName/up:fullName ?proteinLabel .\n",
    "    #     ?diseaseAnnotation up:disease ?diseaseURI .\n",
    "    #     ?diseaseURI skos:prefLabel ?diseaseLabel .\n",
    "    #     FILTER(CONTAINS(LCASE(?diseaseLabel), \"alzheimer\"))\n",
    "    # } LIMIT 10\"\"\",\n",
    "    #     },\n",
    "\n",
    "    #     {\n",
    "    #         # Always failing the connection between OMA and Bgee\n",
    "    #         \"question\": \"For protein Q9Y2T1, what are the orthologs in rats for genes expressed in the brain? Return ?ratProtein ?ratUniProtXref\",\n",
    "    #         \"endpoint\": \"https://sparql.omabrowser.org/sparql/\",\n",
    "    #         \"query\": \"\"\"PREFIX up: <http://purl.uniprot.org/core/>\n",
    "    # PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
    "    # PREFIX orth: <http://purl.org/net/orth#>\n",
    "    # PREFIX lscr: <http://purl.org/lscr#>\n",
    "    # PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
    "    # SELECT DISTINCT ?ratProtein ?ratUniProtXref\n",
    "    # WHERE {\n",
    "    #     ?cluster a orth:OrthologsCluster.\n",
    "    #     ?cluster orth:hasHomologousMember ?node1.\n",
    "    #     ?cluster orth:hasHomologousMember ?node2.\n",
    "    #     ?node1 orth:hasHomologousMember* ?humanProtein.\n",
    "    #     ?node2 orth:hasHomologousMember* ?ratProtein.\n",
    "    #     ?humanProtein a orth:Protein;\n",
    "    #         lscr:xrefUniprot <http://purl.uniprot.org/uniprot/Q9Y2T1>.\n",
    "    #     ?ratProtein a orth:Protein;\n",
    "    #         orth:organism/obo:RO_0002162/up:scientificName 'Rattus norvegicus';\n",
    "    #         lscr:xrefUniprot ?ratUniProtXref.\n",
    "    #     FILTER(?node1 != ?node2)\n",
    "    # }\"\"\",\n",
    "    #     },\n",
    "\n",
    "    #     {\n",
    "    #         # It's making up URIs for photosynthesis instead of filtering on label\n",
    "    #         \"question\": \"Find all proteins that have a known 3D structure and are involved in photosynthesis. Return ?protein, ?structure, ?keywordLabel, limit to 20\",\n",
    "    #         \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "    #         \"query\": \"\"\"PREFIX up: <http://purl.uniprot.org/core/>\n",
    "    # PREFIX keywords: <http://purl.uniprot.org/keywords/>\n",
    "    # PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
    "    # SELECT DISTINCT ?protein ?structure ?keywordLabel\n",
    "    # WHERE {\n",
    "    #     ?protein a up:Protein ;\n",
    "    #         up:classifiedWith/rdfs:label ?keywordLabel ;\n",
    "    #         rdfs:seeAlso ?structure .\n",
    "    #     ?structure up:database ?db .\n",
    "    #     ?db up:category '3D structure databases' .\n",
    "    #     FILTER(LCASE(?keywordLabel) = \"photosynthesis\")\n",
    "    # } LIMIT 20\"\"\",\n",
    "    #     },\n",
    "\n",
    "    # {\n",
    "    #     # Asking for \"full name\" will cause the RAG to fail, but validation don't catch it yet because VoID don't register the class when we have\n",
    "    #     # we need to wait for LinkSet to be added to VoID in UniProt\n",
    "    #     # Why not using \"up:Protein up:enzyme ??\"... Also up:enzymeClass never shows up in the relevant documents\n",
    "    #         \"question\": \"How can I get a list of proteins that are enzymes with their enzyme class? Return the ?proteinURI, ?enzymeClassURI, ?enzymeClassName, limit to 20\",\n",
    "    #         \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "    #         \"query\": \"\"\"PREFIX skos: <http://www.w3.org/2004/02/skos/core#>\n",
    "    #     PREFIX up: <http://purl.uniprot.org/core/>\n",
    "    #     PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
    "    #     SELECT ?proteinURI ?enzymeClassURI ?enzymeClassName\n",
    "    #     WHERE {\n",
    "    #        ?proteinURI a up:Protein ;\n",
    "    #                    up:enzyme ?enzymeClassURI .\n",
    "    #        ?enzymeClassURI skos:prefLabel ?enzymeClassName .\n",
    "    #     #   ?protein a up:Protein ;\n",
    "    #     #            up:annotation ?enzymeAnnotation .\n",
    "    #     #   ?enzymeAnnotation a up:Catalytic_Activity_Annotation ;\n",
    "    #     #                     up:catalyticActivity ?enzymeActivity .\n",
    "    #     #   ?enzymeActivity up:enzymeClass ?enzymeClassURI .\n",
    "    #     #   ?enzymeClassURI skos:prefLabel ?enzymeClassName .\n",
    "    #     } LIMIT 20\"\"\",\n",
    "    # },\n",
    "\n",
    "    # {\n",
    "    #     \"question\": \"Which are the human genes associated with cancer (which have cancer in their disease label) and their orthologs expressed in the rat brain? Return the disease label, human gene URI, human gene HGNC symbol, ortholog rat gene URI\",\n",
    "    #     \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "    #     \"query\": \"\"\"PREFIX skos: <http://www.w3.org/2004/02/skos/core#>\n",
    "    # PREFIX up:<http://purl.uniprot.org/core/>\n",
    "    # PREFIX taxon:<http://purl.uniprot.org/taxonomy/>\n",
    "    # PREFIX rdfs:<http://www.w3.org/2000/01/rdf-schema#>\n",
    "    # PREFIX orth:<http://purl.org/net/orth#>\n",
    "    # PREFIX dcterms:<http://purl.org/dc/terms/>\n",
    "    # PREFIX obo:<http://purl.obolibrary.org/obo/>\n",
    "    # PREFIX lscr:<http://purl.org/lscr#>\n",
    "    # PREFIX genex:<http://purl.org/genex#>\n",
    "    # PREFIX sio: <http://semanticscience.org/resource/>\n",
    "    # SELECT DISTINCT ?diseaseLabel ?humanProtein ?hgncSymbol ?orthologRatProtein ?orthologRatGene\n",
    "    # WHERE {\n",
    "    # SERVICE <https://sparql.uniprot.org/sparql> {\n",
    "    #     SELECT DISTINCT * WHERE {\n",
    "    #     ?humanProtein a up:Protein ;\n",
    "    #         up:organism/up:scientificName 'Homo sapiens' ;\n",
    "    #         up:annotation ?annotation ;\n",
    "    #         rdfs:seeAlso ?hgnc .\n",
    "    #     ?hgnc up:database <http://purl.uniprot.org/database/HGNC> ;\n",
    "    #         rdfs:comment ?hgncSymbol .\n",
    "    #     ?annotation a up:Disease_Annotation ;\n",
    "    #             up:disease ?disease .\n",
    "    #     ?disease skos:prefLabel ?diseaseLabel.\n",
    "    #     FILTER CONTAINS (?diseaseLabel, \"cancer\")\n",
    "    #     }\n",
    "    # }\n",
    "    # SERVICE <https://sparql.omabrowser.org/sparql/> {\n",
    "    #     SELECT ?humanProtein ?orthologRatProtein ?orthologRatGene WHERE {\n",
    "    #     ?humanProteinOma a orth:Protein ;\n",
    "    #             lscr:xrefUniprot ?humanProtein .\n",
    "    #     ?orthologRatProtein a orth:Protein ;\n",
    "    #         sio:SIO_010079 ?orthologRatGene ;\n",
    "    #         orth:organism/obo:RO_0002162/up:scientificName 'Rattus norvegicus' .\n",
    "    #     ?cluster a orth:OrthologsCluster .\n",
    "    #     ?cluster orth:hasHomologousMember ?node1 .\n",
    "    #     ?cluster orth:hasHomologousMember ?node2 .\n",
    "    #     ?node1 orth:hasHomologousMember* ?humanProteinOma .\n",
    "    #     ?node2 orth:hasHomologousMember* ?orthologRatProtein .\n",
    "    #     FILTER(?node1 != ?node2)\n",
    "    #     }\n",
    "    # }\n",
    "    # SERVICE <https://www.bgee.org/sparql/> {\n",
    "    #     ?orthologRatGene genex:isExpressedIn ?anatEntity ;\n",
    "    #         orth:organism ?ratOrganism .\n",
    "    #     ?anatEntity rdfs:label 'brain' .\n",
    "    #     ?ratOrganism obo:RO_0002162 taxon:10116 .\n",
    "    # }\n",
    "    # }\"\"\",\n",
    "    # },\n",
    "]\n",
    "\n",
    "\n",
    "def result_sets_are_same(gen_set, ref_set) -> bool:\n",
    "    gen_set, ref_set = list(gen_set), list(ref_set)\n",
    "    for item in gen_set:\n",
    "        if item not in ref_set:\n",
    "            print(f\"Missing from reference: {item}\")\n",
    "            return False\n",
    "    for item in ref_set:\n",
    "        if item not in gen_set:\n",
    "            print(f\"Missing from generated: {item}\")\n",
    "            return False\n",
    "    return True\n",
    "\n",
    "\n",
    "# QLEVER_UNIPROT = \"https://qlever.cs.uni-freiburg.de/api/uniprot\"\n",
    "\n",
    "# Price per million tokens, open source models based on fireworks.io pricing\n",
    "# https://openai.com/api/pricing/\n",
    "# https://fireworks.ai/pricing\n",
    "models = {\n",
    "    \"Llama3.1 8B\": {\n",
    "        \"id\": \"hf:meta-llama/Meta-Llama-3.1-8B-Instruct\",\n",
    "        \"price_input\": 0.2,\n",
    "        \"price_output\": 0.2,\n",
    "    },\n",
    "    \"Mixtral 8x22B\": {\n",
    "        \"id\": \"hf:mistralai/Mixtral-8x22B-Instruct-v0.1\",\n",
    "        \"price_input\": 1.20,\n",
    "        \"price_output\": 1.20,\n",
    "    },\n",
    "    \"gpt-4o\": {\n",
    "        \"id\": \"gpt-4o\",\n",
    "        \"price_input\": 5,\n",
    "        \"price_output\": 15,\n",
    "    },\n",
    "    \"gpt-4o-mini\": {\n",
    "        \"id\": \"gpt-4o-mini\",\n",
    "        \"price_input\": 0.15,\n",
    "        \"price_output\": 0.6,\n",
    "    },\n",
    "}\n",
    "\n",
    "\n",
    "def answer_no_rag(question: str, model: str):\n",
    "    client = get_llm_client(model)\n",
    "    response = client.chat.completions.create(\n",
    "        model=model,\n",
    "        messages=[\n",
    "            {\"role\": \"system\", \"content\": settings.system_prompt},\n",
    "            {\"role\": \"user\", \"content\": question},\n",
    "        ],\n",
    "        stream=False,\n",
    "        temperature=0,\n",
    "    )\n",
    "    return response.to_dict()\n",
    "\n",
    "\n",
    "def answer_rag_without_validation(question: str, model: str):\n",
    "    response = requests.post(\n",
    "        \"http://localhost:8000/chat\",\n",
    "        json={\n",
    "            \"model\": model,\n",
    "            \"messages\": [{\"role\": \"user\", \"content\": question}],\n",
    "            \"stream\": False,\n",
    "            \"api_key\": expasy_api_key,\n",
    "            \"validate_output\": False,\n",
    "        },\n",
    "        timeout=60,\n",
    "    )\n",
    "    return response.json()\n",
    "\n",
    "\n",
    "def answer_rag_with_validation(question: str, model: str):\n",
    "    response = requests.post(\n",
    "        \"http://localhost:8000/chat\",\n",
    "        json={\n",
    "            \"model\": model,\n",
    "            \"messages\": [{\"role\": \"user\", \"content\": question}],\n",
    "            \"stream\": False,\n",
    "            \"api_key\": expasy_api_key,\n",
    "        },\n",
    "        timeout=60,\n",
    "    )\n",
    "    return response.json()\n",
    "\n",
    "\n",
    "# list_of_approaches = [\n",
    "#     answer_no_rag,\n",
    "#     answer_rag_without_validation,\n",
    "#     answer_rag_with_validation,\n",
    "# ]\n",
    "\n",
    "list_of_approaches = {\n",
    "    \"No RAG\": answer_no_rag,\n",
    "    \"RAG without validation\": answer_rag_without_validation,\n",
    "    \"RAG with validation\": answer_rag_with_validation,\n",
    "}\n",
    "\n",
    "results_data = {\n",
    "    \"Model\": [],\n",
    "    \"RAG Approach\": [],\n",
    "    \"Success\": [],\n",
    "    \"Different Results\": [],\n",
    "    \"No Results\": [],\n",
    "    \"Errors\": [],\n",
    "    \"Price\": [],\n",
    "    # 'Precision': [],\n",
    "    # \"Recall\": [],\n",
    "    \"F1\": [],\n",
    "}\n",
    "\n",
    "number_of_tries = 3\n",
    "\n",
    "print(f\"üß™ Testing {len(example_queries)} queries\")\n",
    "\n",
    "# Get results for the reference queries\n",
    "ref_results = []\n",
    "for i, test_query in enumerate(example_queries):\n",
    "    res_ref_finally_pass = False\n",
    "    while not res_ref_finally_pass:\n",
    "        try:\n",
    "            query_start_time = time.time()\n",
    "            res_from_ref = query_sparql(test_query[\"query\"], test_query[\"endpoint\"], timeout=300)[\"results\"][\"bindings\"]\n",
    "            print(f\"Reference query {i} '{test_query['question']}' took {time.time() - query_start_time:.2f} seconds\")\n",
    "            ref_results.append(res_from_ref)\n",
    "            res_ref_finally_pass = True\n",
    "        except Exception as e:\n",
    "            print(f\"Timeout for reference query {i}: {e}, Trying again because we know if should work.\")\n",
    "            res_ref_finally_pass = False\n",
    "    # res_from_ref = query_sparql(test_query[\"query\"], QLEVER_UNIPROT)[\"results\"][\"bindings\"]\n",
    "\n",
    "for model_label, model in models.items():\n",
    "    print(f\"\\nüß† Testing model: {model_label}\\n\")\n",
    "    res = defaultdict(dict)\n",
    "    # e.g. res[\"No RAG\"][\"success\"] += 1\n",
    "    for approach in list_of_approaches:\n",
    "        res[approach] = defaultdict(int)\n",
    "\n",
    "    for query_num, test_query in enumerate(example_queries):\n",
    "        # # Execute the reference query.\n",
    "        # res_ref_finally_pass = False\n",
    "        # while not res_ref_finally_pass:\n",
    "        #     try:\n",
    "        #         res_from_ref = query_sparql(test_query[\"query\"], test_query[\"endpoint\"], timeout=300)[\"results\"][\"bindings\"]\n",
    "        #         res_ref_finally_pass = True\n",
    "        #     except Exception as e:\n",
    "        #         print(f\"Error in reference query: {e}, Trying again because we know if should work :)\")\n",
    "        #         res_ref_finally_pass = False\n",
    "        # # res_from_ref = query_sparql(test_query[\"query\"], test_query[\"endpoint\"], timeout=300)[\"results\"][\"bindings\"]\n",
    "        # # res_from_ref = query_sparql(test_query[\"query\"], QLEVER_UNIPROT)[\"results\"][\"bindings\"]\n",
    "\n",
    "        for approach, approach_func in list_of_approaches.items():\n",
    "            for t in range(number_of_tries):\n",
    "                response = approach_func(test_query[\"question\"], model[\"id\"])\n",
    "                chat_resp_md = response[\"choices\"][0][\"message\"][\"content\"]\n",
    "                res[approach][\"input_tokens\"] += response[\"usage\"][\"prompt_tokens\"]\n",
    "                res[approach][\"output_tokens\"] += response[\"usage\"][\"completion_tokens\"]\n",
    "                # print(response[\"usage\"])\n",
    "                # print(chat_resp_md)\n",
    "                try:\n",
    "                    generated_sparql = extract_sparql_queries(chat_resp_md)[-1]\n",
    "                    if generated_sparql[\"query\"].strip() == test_query[\"query\"].strip():\n",
    "                        print(f\"‚úÖ {t+1}/{number_of_tries} {test_query['question']}. EXACT MATCH\")\n",
    "                        res[approach][\"success\"] += 1\n",
    "                        continue\n",
    "\n",
    "                    # Execute the generated query\n",
    "                    res_from_generated = query_sparql(\n",
    "                        generated_sparql[\"query\"], generated_sparql[\"endpoint_url\"], timeout=300\n",
    "                    )[\"results\"][\"bindings\"]\n",
    "                    # res_from_generated = query_sparql(generated_sparql[\"query\"], QLEVER_UNIPROT)[\"results\"][\"bindings\"]\n",
    "\n",
    "                    if not result_sets_are_same(res_from_generated, ref_results[query_num]):\n",
    "                        if len(res_from_generated) == 0:\n",
    "                            res[approach][\"no_results\"] += 1\n",
    "                        else:\n",
    "                            res[approach][\"different_results\"] += 1\n",
    "                        raise Exception(\n",
    "                            f\"\\nResults mismatch. Ref: {len(ref_results[query_num])} != gen: {len(res_from_generated)}\\n\"\n",
    "                        )\n",
    "                    else:\n",
    "                        print(f\"‚úÖ {t+1}/{number_of_tries} {test_query['question']} = {len(res_from_generated)}\")\n",
    "                        res[approach][\"success\"] += 1\n",
    "\n",
    "                except Exception as e:\n",
    "                    res[approach][\"fail\"] += 1\n",
    "                    if approach != \"No RAG\":\n",
    "                        fail_emoji = \"‚ùå\" if approach == \"RAG with validation\" else \"‚õîÔ∏è\"\n",
    "                        print(f\"{fail_emoji} {t+1}/{number_of_tries} {test_query['question']}\\n{e}\")\n",
    "                        print(generated_sparql[\"query\"])\n",
    "                        print(\"Correct query:\")\n",
    "                        print(test_query[\"query\"])\n",
    "                        print(\"\")\n",
    "\n",
    "        for approach in list_of_approaches:\n",
    "            print(\n",
    "                f\"üéØ {approach} - Success: {res[approach]['success']}, Different results: {res[approach]['different_results']}, No results: {res[approach]['no_results']}, Error: {res[approach]['fail']-res[approach]['no_results']-res[approach]['different_results']}\"\n",
    "            )\n",
    "\n",
    "    for approach, result_row in res.items():\n",
    "        mean_price = (\n",
    "            (result_row[\"input_tokens\"] * model[\"price_input\"] / 1000000)\n",
    "            + (result_row[\"output_tokens\"] * model[\"price_output\"] / 1000000)\n",
    "        ) / (len(example_queries) * number_of_tries)\n",
    "        precision = result_row[\"success\"] / (result_row[\"success\"] + result_row[\"fail\"])\n",
    "        recall = result_row[\"success\"] / (result_row[\"success\"] + result_row[\"fail\"] - result_row[\"different_results\"])\n",
    "        results_data[\"Model\"].append(model_label)\n",
    "        results_data[\"RAG Approach\"].append(approach)\n",
    "        results_data[\"Success\"].append(result_row[\"success\"])\n",
    "        results_data[\"Different Results\"].append(result_row[\"different_results\"])\n",
    "        results_data[\"No Results\"].append(result_row[\"no_results\"])\n",
    "        results_data[\"Errors\"].append(result_row[\"fail\"] - result_row[\"no_results\"] - result_row[\"different_results\"])\n",
    "        results_data[\"Price\"].append(round(mean_price, 5))\n",
    "        # results_data['Precision'].append(precision)\n",
    "        # results_data['Recall'].append(recall)\n",
    "        if precision + recall == 0:\n",
    "            results_data[\"F1\"].append(0)\n",
    "        else:\n",
    "            results_data[\"F1\"].append(round(2 * (precision * recall) / (precision + recall), 2))\n",
    "\n",
    "\n",
    "df = pd.DataFrame(results_data)\n",
    "print(df)\n",
    "print(\"\\n\\n\")\n",
    "print(df.to_csv(\"tests_results.csv\", index=False))\n",
    "\n",
    "# Output Latex table\n",
    "latex_str = \"\"\n",
    "prev_model = list(models.keys())[0]\n",
    "for index, row in df.iterrows():\n",
    "    row_str = \" & \".join([str(item) for item in row])  # Join all values in the row with \" & \"\n",
    "    row_str += \" \\\\\\\\\"\n",
    "    if row[\"Model\"] != prev_model:\n",
    "        latex_str += \"\\\\midrule\\n\"\n",
    "        prev_model = row[\"Model\"]\n",
    "    latex_str += row_str + \"\\n\"\n",
    "with open(\"tests_results_latex.txt\", \"w\") as f:\n",
    "    f.write(latex_str)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "expasy-scripts",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
